cmd/config-host-add.go

```diff
@@ -116,36 +123,72 @@ func addHost(alias string, hostCfgV8 hostConfigV8) {
 	})
 }
 
-func mainConfigHostAdd(ctx *cli.Context) error {
-	checkConfigHostAddSyntax(ctx)
-
-	console.SetColor("HostMessage", color.New(color.FgGreen))
+func newS3Config(args cli.Args) (*Config, *probe.Error) {
+	var (
+		url       = args.Get(0)
+		accessKey = args.Get(1)
+		secretKey = args.Get(2)
+		api       = args.Get(3)
+	)
+
+	// If api is provided we do not auto probe signature, this is
+	// required in situations when signature type is provided by the user.
+	if api != "" {
+		return &Config{
+			AccessKey: accessKey,
+			SecretKey: secretKey,
+			Signature: api,
+			HostURL:   url,
+		}, nil
+	}
 
-	args := ctx.Args()
-	alias := args.Get(0)
-	url := args.Get(1)
-	accessKey := args.Get(2)
-	secretKey := args.Get(3)
-	api := args.Get(4)
+	s3Config := &Config{
+		AccessKey: accessKey,
+		SecretKey: secretKey,
+		Signature: "s3v4",
+		HostURL:   url,
+	}
 
-	parsedURL, _ := urlpkg.Parse(url)
+	s3Client, err := s3New(s3Config)
+	if err != nil {
+		return nil, err.Trace(args...)
+	}
 
-	if isGoogle(parsedURL.Host) {
-		if api == "S3v4" {
-			fatalIf(errInvalidArgument().Trace(api), "Unsupported API signature for url. Please use `mc config host add "+alias+" "+url+" "+accessKey+" "+secretKey+" S3v2` instead.")
+	for content := range s3Client.List(false, false, DirNone) {
```

Comment:
Can we call BucketExists on a random bucketname and not worry about whether it exists or not? but concentrate more or if it returns an error or not.

Reply:
ListBuckets() is better because of multiple reasons.

- There is no requirement on the right error message sent by the server (BucketExists) might bring that, since `mc` needs to work on many clients they might not send the right error.
- ListBuckets() doesn't have any region requirement so doesn't involve any multiple calls underneath.

It would have been nicer if the S3 REST API had a a signature probe call, but we do not have that.  I prefer to keep it ListBuckets().